File: egghead.io/data-structures-and-algorithms-in-js/lesson-05-linked-list-data-structure-in-js/linked-list.js
const Node = require('./node');


class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  push(value) {
    const node = new Node(value);

    if (this.head === null) {
      this.head = node;
      this.tail = node;
      this.size++;

      return node;
    }

    this.tail.next = node;
    this.tail = node;
    this.size++;

    return node;
  }

  pop() {
    if (this.isEmpty()) {
      return null;
    }

    const node = this.tail;

    if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
      this.size--;

      return node;
    }

    let curr = this.head;
    let penultimate;
    while (curr) {
      if (curr.next === this.tail) {
        penultimate = curr;
        break;
      }

      curr = curr.next;
    }

    penultimate.next = null;
    this.tail = penultimate;
    this.size--;

    return node;
  }

  get(index) {
    if (index < 0 || index > this.size - 1) {
      return null;
    }

    if (index === 0) {
      return this.head;
    }

    let curr = this.head;
    let i = 0;

    while (i < index) {
      curr = curr.next;
      i++;
    }

    return curr;
  }

  delete(index) {
    if (index < 0 || index > this.size - 1) {
      return null;
    }

    let curr = this.head;
    
    if (index === 0) {
      this.head = this.head.next;
      this.size--;

      return curr;
    }

    let prev;
    let i = 0;

    while (i < index) {
      prev = curr;
      curr = curr.next;
      i++;
    }

    prev.next = curr.next;

    if (prev.next === null) {
      this.tail = prev;
    }

    this.size--;

    return curr;
  }

  isEmpty() {
    return this.size === 0;
  }

  print() {
    const values = [];
    let curr = this.head;

    while (curr) {
      values.push(curr.value);
      curr = curr.next;
    }

    return values.join(' => ');
  }
}

module.exports = LinkedList;

File: egghead.io/data-structures-and-algorithms-in-js/lesson-05-linked-list-data-structure-in-js/linked_list.py
from node import Node


class LinkedList:
  def __init__(self):
    self.head = None
    self.tail = None
    self.size = 0

  def push(self, value):
    node = Node(value)

    if not self.head:
      self.head = node
      self.tail = node
      self.size += 1

      return node

    self.tail.succ = node
    self.tail = node
    self.size += 1

    return node

  def pop(self):
    if self.is_empty():
      return None

    node = self.tail

    if self.head == self.tail:
      self.head = None
      self.tail = None
      self.size -= 1

      return node

    curr = self.head
    penultimate = None
    while curr:
      if curr.succ == self.tail:
        penultimate = curr
        break

      curr = curr.succ

    penultimate.succ = None
    self.tail = penultimate
    self.size -= 1

    return node

  def get(self, index):
    if index < 0 or index > self.size - 1:
      return None

    if index == 0:
      return self.head

    curr = self.head
    i = 0
    while i < index:
      curr = curr.succ
      i += 1

    return curr

  def delete(self, index):
    if index < 0 or index > self.size - 1:
      return None

    curr = self.head

    if index == 0:
      self.head == self.head.succ
      self.size -= 1

      return curr

    prev = None
    i = 0
    while i < index:
      prev = curr
      curr = curr.succ
      i += 1
    
    prev.succ = curr.succ

    if not prev.succ:
      self.tail = prev

    self.size -= 1

    return curr

  def is_empty(self):
    return self.size == 0

  def print(self):
    values = []
    curr = self.head

    while curr:
      values.append(curr.value)
      curr = curr.succ

    return ' => '.join(map(str, values))